lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -304,6 +327,93 @@ impl Client {
         })
     }
 
+    async fn handle_clipboard(
+        cgo_handle: CgoHandle,
+        write_stream: &mut RdpWriteStream,
+        x224_processor: Arc<Mutex<X224Processor>>,
+        clipboard: &mut Option<String>,
+        clipboard_function: ClipboardFunction,
+    ) -> ClientResult<()> {
+        match clipboard_function {
+            ClipboardFunction::RequestFormatList => {
+                Client::process_cliprdr(x224_processor, write_stream, |c| c.initiate_copy(&[]))
+                    .await?
+            }
+            ClipboardFunction::Update(data) => {
+                clipboard.replace(data);
+                Client::process_cliprdr(x224_processor, write_stream, |c| {
+                    c.initiate_copy(&[ClipboardFormat::new(ClipboardFormatId::CF_UNICODETEXT)])
+                })
+                .await?;
+            }
+            ClipboardFunction::RemoteCopy(data) => {
+                let format = data
+                    .iter()
+                    .find(|cf| {
+                        cf.id() == ClipboardFormatId::CF_UNICODETEXT
+                            || cf.id() == ClipboardFormatId::CF_TEXT
+                    })
+                    .map(ClipboardFormat::id);
+                if let Some(cf) = format {
+                    Client::process_cliprdr(x224_processor, write_stream, move |c| {
+                        c.initiate_paste(cf)
+                    })
+                    .await?;
+                }
+            }
+            ClipboardFunction::FormatDataRequest(req) => {
+                let data = clipboard
+                    .as_ref()
+                    .filter(|_| req == ClipboardFormatId::CF_UNICODETEXT)
+                    .map(|s| util::to_unicode(s.as_str(), true))
+                    .map_or_else(FormatDataResponse::new_error, FormatDataResponse::new_data);
+                Client::process_cliprdr(x224_processor, write_stream, |c| {
+                    c.submit_format_data(data)
+                })
+                .await?;
+            }
+            ClipboardFunction::FormatDataResponse(data) => unsafe {
```

Comment:
Let's move the `unsafe` block as close to the source as possible, rather than making this entire blocking `unsafe`. E.g. remove this and below have:

```rust
.spawn_blocking(move || unsafe {
    handle_remote_copy(cgo_handle, data.as_mut_ptr(), data.len() as u32)
})
```